src/machine/mod.rs

```diff
@@ -673,10 +684,40 @@ impl<Inner> DefaultMachineBuilder<Inner> {
     pub fn build(self) -> DefaultMachine<Inner> {
         DefaultMachine {
             inner: self.inner,
+            pause: Pause::new(),
             instruction_cycle_func: self.instruction_cycle_func,
             debugger: self.debugger,
             syscalls: self.syscalls,
             exit_code: 0,
         }
     }
 }
+
+#[derive(Clone, Default)]
+pub struct Pause {
+    s: Arc<AtomicU8>,
+}
+
+impl Pause {
+    pub fn new() -> Self {
+        Self {
+            s: Arc::new(AtomicU8::new(0)),
+        }
+    }
+
+    pub fn interrupt(&self) {
+        self.s.store(1, Ordering::SeqCst);
+    }
+
+    pub fn has_interrupted(&self) -> bool {
+        self.s.load(Ordering::SeqCst) != 0
+    }
+
+    pub fn get_raw_ptr(&self) -> *mut u8 {
+        &*self.s as *const _ as *mut u8
```

Comment:
Do we need [get_mut](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicU8.html#method.get_mut) here?

Reply:
We cannot borrow data in an `Arc` as mutable.

Reply:
In that sense you will need Mutex to wrap on `AtomicU8`.

What I'm concerned here, is that we are casting a `*AtomicU8` directly to a `*u8`, which means we are relying on the underlying layout that the u8 value lies at the first of the `AtomicU8` structure. This is something that worries me.

Reply:
We can write a test case to guarantee the relationship between AtomicU8 and u8.